evm/src/cpu/kernel/interpreter.rs

```diff
@@ -893,16 +903,25 @@ impl<'a, F: Field> Interpreter<'a, F> {
             .debug_offsets
             .contains(&self.generation_state.registers.program_counter)
         {
-            println!("At {}, stack={:?}", self.offset_name(), self.stack());
+            println!("At {}, stack={:?},", self.offset_name(), self.stack());
         } else if let Some(label) = self.offset_label() {
-            println!("At {label}");
+            println!(
+                "At {label} stack = {:?} is kernel = {:?}",
+                {
+                    let mut stack = self.stack();
+                    stack.reverse();
+                    stack
+                },
+                self.is_kernel()
+            );
+        } else if !self.is_kernel() {
+            println!("User instruction {:?}, stack = {:?}", op, {
+                let mut stack = self.stack();
+                stack.reverse();
+                stack
+            });
         }
 
-        let op = decode(self.generation_state.registers, opcode)
-            // We default to prover inputs, as those are kernel-only instructions that charge nothing.
-            .unwrap_or(Operation::ProverInput);
-        self.generation_state.registers.gas_used += gas_to_charge(op);
```

Comment:
Why has this been moved prior to the opcode execution? (this is different from the actual CPU)

Reply:
Because if you have an invalid exit kernel you will still execute `run_exit_kernel` and set the kernel flag to `false`. Then when you call decode it doesn't return an error.

Reply:
Maybe this is not the best solution, but the problem was that the interpreter was not erroring in invalid EXIT_KERNEL while it should

Reply:
It was with the `undefinedOpcodeFirstByte_d0g0v0_Shanghai` test I think. And the problem is that `run_exit_kernel` doesn't check if it is or not in kernel mode.
 I think the solution should be that the interpreter's `run_opcode` fn does the match with an already decoded opcode like `perfom_op`. I wanted to minimize the changes in the interpreter though

Reply:
Yeah this may be out of scope for this PR, but useful to keep track of such discrepancies nonetheless. Thanks for the info!